src/default-config-backend.cpp

```diff
@@ -9,29 +9,61 @@
 #include <sys/inotify.h>
 #include <unistd.h>
 
-#define INOT_BUF_SIZE (1024 * sizeof(inotify_event))
-static char buf[INOT_BUF_SIZE];
-
+#define INOT_BUF_SIZE (sizeof(inotify_event) + NAME_MAX + 1)
 
 static std::string config_dir, config_file;
 wf::config::config_manager_t *cfg_manager;
 
-static void reload_config(int fd)
+static void readd_watch(int fd)
 {
-    wf::config::load_configuration_options_from_file(*cfg_manager, config_file);
     inotify_add_watch(fd, config_dir.c_str(), IN_CREATE);
     inotify_add_watch(fd, config_file.c_str(), IN_MODIFY);
 }
 
+static void reload_config(int fd)
+{
+    wf::config::load_configuration_options_from_file(*cfg_manager, config_file);
+    readd_watch(fd);
+}
+
 static int handle_config_updated(int fd, uint32_t mask, void *data)
 {
+    if ((mask & WL_EVENT_READABLE) == 0)
+    {
+        return 0;
+    }
+
     LOGD("Reloading configuration file");
 
-    /* read, but don't use */
-    read(fd, buf, INOT_BUF_SIZE);
-    reload_config(fd);
+    char buf[INOT_BUF_SIZE]
+    __attribute__((aligned(alignof(inotify_event))));
+
+    bool should_reload = false;
 
-    wf::get_core().emit_signal("reload-config", nullptr);
+    while (read(fd, buf, INOT_BUF_SIZE) >= sizeof(inotify_event))
```

Comment:
Unfortunately, still wrong. You'd need something like this:

```cpp
read_exact(fd, buf, sizeof(inotify_event));
auto event = reinterpret_cast<inotify_event*>(buf);
read_exact(fd, buf + sizeof(inotify_event), event->len);
```

with read_exact from [here](https://github.com/WayfireWM/wayfire/blob/ipc/src/ipc/ipc.cpp#L175)

Reply:
The reason is: read can be interrupted at any time, so in your loop read might just get 1 byte the first time, but read much more bytes the second time. This would mean that you're reading mis-aligned data.

Reply:
I don't think that would work. According to the manual:

```
       The behavior when the buffer given to read(2) is too small to
       return information about the next event depends on the kernel
       version: in kernels before 2.6.21, read(2) returns 0; since
       kernel 2.6.21, read(2) fails with the error EINVAL.
```

This suggests that reading a small buffer the size of exactly `inotify_event` will not work if the event comes with a name after it. Instead, the `read` will error out with `EINVAL`.

I don't think my implementation is incorrect in that `read` might return an incomplete event that is unaccounted for.

The implementation does seem to not account for multiple `inotify_event`s in the same buffer, however. The C implementation does this by incrementing the buffer pointer by `sizeof(inotify_event) + event->len`.

Reply:
Ah, so it seems that inotify guarantees we'll get a full event when we read from the FD, which is quite nice. I should have read the inotify docs too, not just the ones for read.

```
Each successful read(2) returns a buffer containing one or more of the following structures:
```



> The implementation does seem to not account for multiple `inotify_event`s in the same buffer, however. The C implementation does this by incrementing the buffer pointer by `sizeof(inotify_event) + event->len`.

Would it be hard to handle multiple events here too?

Reply:
I don't think it'll be hard, no. I just have to do it.

Reply:
I've rebased in a fix for this. It's untested though.